### DISCUSSION TITLE
subscriptions not calling source

### DISCUSSION THREAD

**Discussion by korzi27** [mirumee/ariadne#1163] [Q&A]
Hello there,

very similar to [Issue 320](https://github.com/mirumee/ariadne/issues/320) I experienced errors in my productive app that subscriptions are not calling the source definition. Going back to the documentary example i experience the same problem:

```
import asyncio
from ariadne import SubscriptionType, make_executable_schema
from ariadne.asgi import GraphQL
from ariadne.asgi.handlers import GraphQLWSHandler


type_def = """
    type Query {
        _unused: Boolean
    }

    type Subscription {
        counter: Int!
    }
"""

subscription = SubscriptionType()

@subscription.source("counter")
async def counter_generator(obj, info):
    print('Source')
    for i in range(5):
        await asyncio.sleep(1)
        yield i


@subscription.field("counter")
def counter_resolver(count, info):
    print('field')
    return count + 1


schema = make_executable_schema(type_def, subscription)
app = GraphQL(schema, debug=True, websocket_handler = GraphQLWSHandler())
```

Calling with uvicorn app:app

Now i am trying to call the subsricptions with

```
import asyncio
import httpx

async def graphql_subscription(endpoint, subscription_query):
    async with httpx.AsyncClient() as client:
        response = await client.post(endpoint, json={'query': subscription_query})
        print(response.json())
        data = response.json()
        print(data)


graphql_endpoint = 'http://localhost:8000/graphql/'
subscription_query = '''subscription MySubscri{
                        counter}'''


asyncio.run(graphql_subscription(graphql_endpoint, subscription_query))
```

Ending up with the error message `unsupported operand type(s) for +: 'NoneType' and 'int'` as count+1 cannot be computed as count is None as the source is never called. 

python 3.10.5, ariadne 0.22, httpx 0.22.6, uvicorn 0.27.1

Help is pretty much appreciated.

**Comment by rafalp** [comment]
`client.post`

Subscriptions should only work with websockets, so fact that even subscription's resolver is called is a bug in Ariadne.

**Comment by rafalp** [comment]
We'll fix this in Ariadne 0.23 so trying to do a subscription without using a WebSocket will return an error.

For posterity what happened here is there are two query execution strategies in GraphQL: `execute` and `subscribe`. `execute` just calls resolvers in a cascade. `subscribe` calls sources and whenever a source yields a result, calls its resolver.

`subscribe` is only used for connections made over the websocket. What happened there is `execute` was called against `Subscription` fields resolvers like it would be against `Query` or `Mutation` and getting error.
